README.md
## adhocore/phint

Initializes new PHP project with sane defaults using templates.
It scaffolds PHP library &/or project to boost your productivity and save time.

For already existing project, run with `--sync` flag to add missing stuffs, see [phint init](#init).

Once you have files in your `src/` or `lib/` you can run [phint docs](#docs) to generate API like documentation in `.md` format
and [phint test](#test) to generate basic test stubs with all the structures already maintained.

It helps you be even more lazier! **phint** is continuously evolving and the plan is to make it [big](#todo).

[![Latest Version](https://img.shields.io/github/release/adhocore/phint.svg?style=flat-square)](https://github.com/adhocore/phint/releases)
[![Travis Build](https://img.shields.io/travis/adhocore/phint/master.svg?style=flat-square)](https://travis-ci.org/adhocore/phint?branch=master)
[![Scrutinizer CI](https://img.shields.io/scrutinizer/g/adhocore/phint.svg?style=flat-square)](https://scrutinizer-ci.com/g/adhocore/phint/?branch=master)
[![Codecov branch](https://img.shields.io/codecov/c/github/adhocore/phint/master.svg?style=flat-square)](https://codecov.io/gh/adhocore/phint)
[![StyleCI](https://styleci.io/repos/108550679/shield)](https://styleci.io/repos/108550679)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)

![Phint Preview](https://imgur.com/F6PkX9Z.png "Phint Preview")

[Installation](#installation) &middot; [Features](#features) &middot; [Autocompletion](#autocompletion) &middot; [Usage](#usage) &middot; [phint init](#init) &middot; [phint update](#update) &middot; [phint docs](#docs) &middot; [phint test](#test) &middot; [Templating](#templating)

> Phint is powered by [adhocore/cli](https://github.com/adhocore/php-cli)

## Installation

> Requires PHP7.

### Manual

Download `phint.phar` from [latest release](https://github.com/adhocore/phint/releases/latest).
And use it like so `php /path/to/phint.phar [opts] [args]`. Hmm not cool. See Command section below.

### Command

```bash
# get latest version (you need `jq`)
LATEST_PHINT=`curl --silent "https://api.github.com/repos/adhocore/phint/releases/latest" | jq -r .tag_name`

# download latest phint
curl -sSLo ~/phint.phar "https://github.com/adhocore/phint/releases/download/$LATEST_PHINT/phint.phar"

# make executable
chmod +x ~/phint.phar
sudo ln -s ~/phint.phar /usr/local/bin/phint

# check
phint --help
```

## Features

- generate dot files the likes of `.gitignore, .travis.yml, . editorconfig` etc
- generate `LICENSE`, `README.md`, `composer.json`
- generate `CHANGELOG.md` stub, `CONTRIBUTING.md` guide, `ISSUE_TEMPLATE.md` and `PULL_REQUEST_TEMPLATE.md`
- generate binaries if any
- git init
- interactively ask and install all the dev and prod deps
- generate `phpunit.xml`, test `bootstrap.php`
- generate test stubs for all classes/methods corresponding to `src` (`phint test`)
- generate docs for all public class/methods
- export templates to chosen path so it can be customized (`phint export`)
- use custom templates from a path specified by user
- update its own self (`phint update`)

## Autocompletion

The phint commands and options can be [autocompleted](https://github.com/adhocore/php-cli#autocompletion) if you use zsh shell with oh-my-zsh.

Setting up auto complete:

```sh
mkdir -p ~/.oh-my-zsh/custom/plugins/ahccli && cd ~/.oh-my-zsh/custom/plugins/ahccli

[ -f ./ahccli.plugin.zsh ] || curl -sSLo ./ahccli.plugin.zsh https://raw.githubusercontent.com/adhocore/php-cli/master/ahccli.plugin.zsh

echo compdef _ahccli phint >> ./ahccli.plugin.zsh

chmod +x ./ahccli.plugin.zsh && source ./ahccli.plugin.zsh && cd -
```

Dont forget to [add](https://github.com/adhocore/php-cli#load-ahccli-plugin) `ahccli` into `plugins=(... ...)` list in `~/.zshrc` file.

## Usage

It can be used to quickly spin off new  project containing all basic and default stuffs. The quick steps are as follows:

```bash
# See options/arguments
phint init --help

# OR (shortcut)
phint i -h

# Below command inits a brand new PHP project in `project-name` folder in current dir
# Missing arguments are interactively collected
phint init project-name

# You can also use config file (with json) to read option values from
phint init project-name --config phint.json
```

## Commands

Each of the commands below should be used like so:
```sh
cd /path/to/project
phint <command> [--options] [args]
```

### init

> alias i

Create and Scaffold a bare new PHP project.

***Parameters:***

Dont be intimidated by long list of parameters, you are not required to enter any of them
as arguments as they are interactively collected when required.

Also check [config](#example-config) on how to create a reusable json config so you can use `phint` like a *pro*.

```
Arguments:
  <project>  The project name without slashes

Options:
  [-b, --bin...]            Executable binaries
  [-c, --no-codecov]        Disable codecov
  [-C, --config]            JSON filepath to read config from
  [-d, --descr]             Project description
  [-D, --dev...]            Developer packages
  [-e, --email]             Vendor email
  [-f, --force]             Run even if the project exists
  [-G, --gh-template]       Use `.github/` as template path
                            By default uses `docs/`
  [-h, --help]              Show help
  [-w, --keywords...]       Project Keywords
  [-L, --license]           License (m: MIT, g: GNULGPL, a: Apache2, b: BSDSimple, i: ISC, w: WTFPL)
  [-n, --name]              Vendor full name
  [-N, --namespace]         Root namespace (use `/` separator)
  [-g, --package]           Packagist name (Without vendor handle)
  [-p, --path]              The project path (Auto resolved)
  [-P, --php]               Minimum PHP version
  [-R, --req...]            Required packages
  [-s, --no-scrutinizer]    Disable scrutinizer
  [-l, --no-styleci]        Disable StyleCI
  [-S, --sync]              Only create missing files
                            Use with caution, take backup if needed
  [-t, --no-travis]         Disable travis
  [-T, --type]              Project type
  [-u, --username]          Vendor handle/username
  [-z, --using]             Reference package
  [-y, --year]              License Year

Usage Examples:
  phint init <project> --force --descr "Awesome project" --name "YourName" --email you@domain.com
  phint init <project> --using laravel/lumen --namespace Project/Api --type project</comment>
  phint init <project> --php 7.0 --config /path/to/json --dev mockery/mockery --req adhocore/cli
```

### Example config

Parameters sent via command args will have higher precedence than values from config file (`-C --config`).

What can you put in config? Anything but we suggest you put only known options (check `$ phint init --help`)

```json
{
  "type": "library",
  "namespace": "Ahc",
  "username": "adhocore",
  "name": "Jitendra Adhikari",
  "email": "jiten.adhikary@gmail.com",
  "php": "7.0",
  "codecov": false,
  "...": "..."
}
```

---
## update

> alias u

Update Phint to lastest version or rollback to earlier locally installed version.

***Parameters:***

```
Options:
  [-h, --help]         Show help
  [-r, --rollback]     Rollback to earlier version

Usage Examples:
  phint update        Updates to latest version
  phint u             Also updates to latest version
  phint update -r     Rolls back to prev version
  phint u --rollback  Also rolls back to prev version
```

---
## docs

> alias d

Generate docs (`.md`) for all public classes and methods from their docblocks.

Ideally you would run it on existing project **or** after you create/update `src/` files.

***Parameters:***

```
Options:
  [-a, --with-abstract]    Create docs for abstract/interface class
  [-h, --help]             Show help
  [-o, --output]           Output file (default README.md). For old project you should use something else
                           (OR mark region with <!-- DOCS START --> and <!-- DOCS END --> to inject docs)

Usage Examples:
  phint docs               If there is `<!-- DOCS START -->` and `<!-- DOCS END -->` region
                           Injects new doc in between them otherwise appends to bottom
  phint d -o docs/api.md   Writes to docs/api.md (Same rule applies regarding inject/append)
```

### Sample docs

***PHP code***

```php
namespace Abc;

/**
 * This is dummy class.
 *
 * It does nothing as of now.
 * Maybe you could fix it?
 */
class Dummy
{
    /**
     * Alpha beta.
     *
     * Example:
     *
     * <code>
     * $dummy = new Dummy;
     * $dummy->alpha('john', true);
     * // '...'
     * </code>
     *
     * @param string $name
     * @param bool   $flag
     *
     * @return string|null
     */
    public function alpha($name, $flag)
    {
        //
    }
}
```

***Generated Markdown***

```md
## Dummy

```php
use Abc\Dummy;
\```

> This is dummy class.

It does nothing as of now.
Maybe you could fix it?

### alpha()

> Alpha beta.

```php
alpha(string $name, bool $flag): string|null
\```

Example:

```php
$dummy = new Dummy;
$dummy->alpha('john', true);
// '...'
\```
```

***Preview***

## Dummy

```php
use Ahc\Dummy;
```

> This is dummy class.

It does nothing as of now.
Maybe you could fix it?

### alpha()

> Alpha beta.

```php
alpha(string $name, bool $flag): string|null
```

Example:

```php
$dummy = new Dummy;
$dummy->alpha('john', true);
// '...'
```

---
## test

> alias t

Generate test files with proper classes and test methods analogous to their source counterparts.
If a test class already exists, it is skipped. In future we may append test stubs for new methods.

Ideally you would run it on existing project **or** after you create/update `src/` files.

***Parameters:***

```
Options:
  [-a, --with-abstract]    Create stub for abstract/interface class
  [-h, --help]             Show help
  [-n, --naming]           Test method naming format
                           (t: testMethod | m: test_method | i: it_tests_)
  [-p, --phpunit]          Base PHPUnit class to extend from
  [-s, --no-setup]         Dont add setup method
  [-t, --no-teardown]      Dont add teardown method

Usage Examples:
  phint test -n i        With `it_` naming
  phint t --no-teardown  Without `tearDown()`
  phint test -a          With stubs for abstract/interface
```

### Sample test

Generated `tests/Dummy.php` for [Abc\\Dummy](#sample-docs) above:

```php
<?php

namespace Abc\Test;

use Abc\Dummy;
use PHPUnit\Framework\TestCase as TestCase;

/**
 * Auto generated by `phint test`.
 */
class DummyTest extends TestCase
{
    /**
     * @var Dummy
     */
    protected $dummy;

    public function setUp()
    {
        parent::setUp();

        $this->dummy = new Dummy;
    }

    public function testAlpha()
    {
        $actual = $this->dummy->alpha();

        // $this->assertSame('', $actual);
    }
}
```

---
## Templating

> `phint export --to ~/myphint`

So you would like to have your own templates and customize `phint` to your taste!

First you need to create a directory root (of any name, eg: `myphint`) with structure that looks like:

```tree
myphint
├── CHANGELOG.md.twig
├── composer.json.twig
├── CONTRIBUTING.md.twig
├── docs
│   ├── docs.twig
│   ├── ISSUE_TEMPLATE.md.twig
│   └── PULL_REQUEST_TEMPLATE.md.twig
├── .editorconfig.twig
├── .env.example.twig
├── .gitignore.twig
├── LICENSE.twig
├── package.json.twig
├── phpunit.xml.dist.twig
├── README.md.twig
├── tests
│   ├── bootstrap.php.twig
│   └── test.twig
└── .travis.yml.twig
```

Note that you dont need to have all the files there in new directory just pick the ones you would like to customize and start hacking.

Luckily you **dont** have to create these templates yourself, just run `phint export --to ~/myphint`!

**Pro Tip**
You can actually introduce any new template as long as their extension is `.twig`.
Such templates are *only* used by `phint init` command. Check [Template variables](#template-variables).

After you are done customizing these templates you can use them in each of the *phint* commands like so

```sh
phint init project --template ~/myphint
phint docs --template ~/myphint
phint test --template ~/myphint
```

The short option name for `--template` is `-x`.

#### Template variables

Here's what parameters these templates would receive when run:

- ***[docs/docs.twig](https://github.com/adhocore/phint/blob/master/resources/docs/docs.twig):*** classes [metadata](#class-metadata) and [docs parameters](#docs)
- ***[tests/test.twig](https://github.com/adhocore/phint/blob/master/resources/tests/test.twig):*** class [metadata](#class-metadata) and [test parameters](#test)
- ***Everything else:*** [init parameters](#init)

***Metadata***

- The `docs` and `test` commands read and use source files metadata.
- The `docs.twig` template recieves metadata collection of all classes at once.
- The `test.twig` template recieves metadata unit of one class at a time.

### Class metadata

Example metadata for [Abc\\Dummy](#sample-docs) above:

```php
[
  'namespace'   => 'Abc',
  'classFqcn'   => 'Abc\\Dummy',
  'classPath'   => '/home/user/projects/src/Dummy.php',
  'name'        => 'Dummy',
  'className'   => 'Dummy',
  'isTrait'     => false,
  'isAbstract'  => false,
  'isInterface' => false,
  'newable'     => true,
  'title'       => 'This is dummy class.',
  'texts'       => [
    'It does nothing as of now.',
    'Maybe you could fix it?',
  ],
  'methods' => [
    'alpha' => [
      'name'       => 'alpha',
      'inClass'    => 'Abc\\Dummy',
      'isStatic'   => false,
      'isFinal'    => false,
      'isPublic'   => true,
      'isAbstract' => false,
      'maybeMagic' => false,
      'title'      => 'Alpha beta.',
      'texts'      => [
        'Example:',
        '<code>',
        '$dummy = new Dummy;',
        '$dummy->alpha(\'john\', true);',
        '// \'...\'',
        '</code>',
      ],
      'return' => 'string|null',
      'params' => [
        'string $name',
        'bool $flag',
      ],
    ],
    // more methods ...
  ],
];
```

## Todo

Including but not limited to:

- [x] README.md/Docs generator
- [x] Test files generator
- [x] Support user templates
- [ ] Test stubs for new methods

## License

> &copy; 2017-2020, [Jitendra Adhikari](https://github.com/adhocore) | [MIT](./LICENSE)

### Credits

This library is release managed by [please](https://github.com/adhocore/please).


CHANGELOG.md
## [0.6.1](https://github.com/adhocore/phint/releases/tag/0.6.1) (2020-01-03)

### Miscellaneous
- **Composer**: Bump adhocore/cli (Jitendra Adhikari) [_b48e4d3_](https://github.com/adhocore/phint/commit/b48e4d3)


## [0.6.0](https://github.com/adhocore/phint/releases/tag/0.6.0) (2020-01-03)

### Miscellaneous
- **Composer**: Bump cli and json-comment (Jitendra Adhikari) [_e872c21_](https://github.com/adhocore/phint/commit/e872c21)


## [v0.5.4](https://github.com/adhocore/phint/releases/tag/v0.5.4) (2020-01-03)

### Miscellaneous
- **Composer**: Bump adhocore/cli (Jitendra Adhikari) [_862cf7a_](https://github.com/adhocore/phint/commit/862cf7a)

### Documentations
- Add credits, update license year (Jitendra Adhikari) [_2dfdf67_](https://github.com/adhocore/phint/commit/2dfdf67)


## [v0.5.3](https://github.com/adhocore/phint/releases/tag/v0.5.3) (2019-12-31)

### Bug Fixes
- **Phpunit**: Rm xml.syntaxCheck (Jitendra Adhikari) [_50b872b_](https://github.com/adhocore/phint/commit/50b872b)

### Miscellaneous
- **Template**: Update test stuffs (Jitendra Adhikari) [_d8ccb9b_](https://github.com/adhocore/phint/commit/d8ccb9b)
- **Travis**: Script (Jitendra Adhikari) [_6322753_](https://github.com/adhocore/phint/commit/6322753)
- **Composer**: Tweak script.test (Jitendra Adhikari) [_fd12251_](https://github.com/adhocore/phint/commit/fd12251)
- **Deps**: Bump cli (Jitendra Adhikari) [_c20ca7e_](https://github.com/adhocore/phint/commit/c20ca7e)

### Builds
- **Deps**: Update adhocore/cli requirement from ^0.6 to ^0.6 || ^0.7 (dependabot-preview[bot]) [_31b6101_](https://github.com/adhocore/phint/commit/31b6101)


## [v0.5.2](https://github.com/adhocore/phint/releases/tag/v0.5.2) (2019-12-28)

### Miscellaneous
- Bump adhocore/cli (Jitendra Adhikari) [_dce602c_](https://github.com/adhocore/phint/commit/dce602c)

### Builds
- **Deps**: Update adhocore/cli requirement from ^0.5 to ^0.5 || ^0.6 (dependabot-preview[bot]) [_a35483a_](https://github.com/adhocore/phint/commit/a35483a)


## [v0.5.1](https://github.com/adhocore/phint/releases/tag/v0.5.1) (2019-12-16)

### Miscellaneous
- **Template**: Bootstrapped by phint (Jitendra Adhikari) [_7f33932_](https://github.com/adhocore/phint/commit/7f33932)
- Bump adhocore/cli (Jitendra Adhikari) [_05791b3_](https://github.com/adhocore/phint/commit/05791b3)

### Builds
- **Deps**: Update adhocore/cli requirement from ^0.3 to ^0.3 || ^0.5 (dependabot-preview[bot]) [_69d2322_](https://github.com/adhocore/phint/commit/69d2322)


## [v0.5.0] 2019-03-05 14:03:22 UTC

- fix(basecmd): autoload classes (Jitendra Adhikari)
- feat(util.composer): add force autoload (Jitendra Adhikari)

## [v0.4.1] 2019-02-09 15:02:20 UTC

- [f303d13](https://github.com/adhocore/phint/commit/f303d13) build(travis): no nightly (Jitendra Adhikari)
- [b8d3bc3](https://github.com/adhocore/phint/commit/b8d3bc3) build(travis): php 7.3 (Jitendra Adhikari)
- [9fa7a31](https://github.com/adhocore/phint/commit/9fa7a31) chore: also list php 7.3 (Jitendra Adhikari)

## [v0.4.0] 2018-09-23 03:09:47 UTC

- [fce6e4e](https://github.com/adhocore/phint/commit/fce6e4e) fix(util.executable): log if only writable (Jitendra Adhikari)
- [b9d8c09](https://github.com/adhocore/phint/commit/b9d8c09) chore: composer remove unused deps (Jitendra Adhikari)
- [e178127](https://github.com/adhocore/phint/commit/e178127) refactor(util.executable): log output, cleanup unused imports (Jitendra Adhikari)
- [c53e469](https://github.com/adhocore/phint/commit/c53e469) feat(util.executable): add windows binary finder (Jitendra Adhikari)
- [b51feee](https://github.com/adhocore/phint/commit/b51feee) refactor(util.executable): use ahc/cli/shell and findBinary (Jitendra Adhikari)
- [56747a3](https://github.com/adhocore/phint/commit/56747a3) chore: composer deps bump (Jitendra Adhikari)

## [v0.3.6] 2018-09-20 15:09:16 UTC

- [0485057](https://github.com/adhocore/phint/commit/0485057) fix(util.meta): param must be array (Jitendra Adhikari)
- [3f3fc5e](https://github.com/adhocore/phint/commit/3f3fc5e) test: fix meta test (Jitendra Adhikari)
- [bb27989](https://github.com/adhocore/phint/commit/bb27989) test: fix git test (Jitendra Adhikari)
- [ef48c80](https://github.com/adhocore/phint/commit/ef48c80) chore(docs.tmpl): make collaspible. add throws. fix return/param (Jitendra Adhikari)
- [a8898a6](https://github.com/adhocore/phint/commit/a8898a6) refactor(util.metadata): add throws, improve return/params (Jitendra Adhikari)
- [72bc7fa](https://github.com/adhocore/phint/commit/72bc7fa) fix(twig.generator): fix call() fn (Jitendra Adhikari)
- [92230ae](https://github.com/adhocore/phint/commit/92230ae) fix(initcmd): add missing path arg in projectExists() (Jitendra Adhikari)
- [a443885](https://github.com/adhocore/phint/commit/a443885) docs: update autocomplete section (Jitendra Adhikari)
- [79029f9](https://github.com/adhocore/phint/commit/79029f9) Intialize Util Composer, Git and Metadata tests (peter279k)

## [v0.3.5] 2018-09-01 12:09:26 UTC

- [376fea9](https://github.com/adhocore/phint/commit/376fea9) chore: support wtfpl license [closes #52] (Jitendra Adhikari)
- [487681b](https://github.com/adhocore/phint/commit/487681b) chore: add wtfpl license support (Jitendra Adhikari)
- [a0c040b](https://github.com/adhocore/phint/commit/a0c040b) docs: template with anchor and jumps [closes #47] (Jitendra Adhikari)
- [f5f2242](https://github.com/adhocore/phint/commit/f5f2242) Add generateTest and generateDoc tests (peter279k)

## [v0.3.4] 2018-08-31 15:08:19 UTC

- [5003ec4](https://github.com/adhocore/phint/commit/5003ec4) fix(initcmd): donot prompt --template (Jitendra Adhikari)
- [1356bc3](https://github.com/adhocore/phint/commit/1356bc3) fix(basecmd): realpath fails on phar (Jitendra Adhikari)
- [03616f3](https://github.com/adhocore/phint/commit/03616f3) docs: reorder test section, add sample test (Jitendra Adhikari)

## [v0.3.3] 2018-08-31 12:08:46 UTC


src/Console/InitCommand.php
<?php

/*
 * This file is part of the PHINT package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

namespace Ahc\Phint\Console;

use Ahc\Cli\Exception\InvalidArgumentException;
use Ahc\Cli\Input\Command;
use Ahc\Cli\IO\Interactor;
use Ahc\Phint\Generator\CollisionHandler;
use Ahc\Phint\Generator\TwigGenerator;
use Ahc\Phint\Util\Inflector;

class InitCommand extends BaseCommand
{
    /** @var string Command name */
    protected $_name = 'init';

    /** @var string Command description */
    protected $_desc = 'Create and Scaffold a bare new PHP project';

    /**
     * Configure the command options/arguments.
     *
     * @return void
     */
    protected function onConstruct()
    {
        $this
            ->argument('<project>', 'The project name without slashes')
            ->option('-T --type', 'Project type (project | library | composer-plugin)')
            ->option('-n --name', 'Vendor full name', null, $this->_git->getConfig('user.name'))
            ->option('-e --email', 'Vendor email', null, $this->_git->getConfig('user.email'))
            ->option('-u --username', 'Vendor handle/username')
            ->option('-N --namespace', 'Root namespace (use `/` separator)')
            ->option('-P --php', 'Minimum PHP version', 'floatval')
            ->option('-p --path', 'The project path (Auto resolved)')
            ->option('-S --sync', "Only create missing files\nUse with caution, take backup if needed", null, false)
            ->option('-f --force', "Run even if the project exists\nUse with caution, take backup if needed", null, false)
            ->option('-g --package', 'Packagist name (Without vendor handle)')
            ->option('-x --template', "User supplied template path\nIt has higher precedence than inbuilt templates")
            ->option('-d --descr', 'Project description')
            ->option('-w --keywords [words...]', 'Project Keywords')
            ->option('-y --year', 'License Year', null, date('Y'))
            ->option('-b --bin [binaries...]', 'Executable binaries')
            ->option('-z --using', 'Reference package (should be known to composer)')
            ->option('-C --config', 'JSON filepath to read config from')
            ->option('-G --gh-template', "Use `.github/` as template path\nBy default uses `docs/`", null, false)
            ->option('-R --req [pkgs...]', 'Required packages')
            ->option('-D --dev [pkgs...]', 'Developer packages')
            ->option('-t --no-travis', 'Disable travis')
            ->option('-c --no-codecov', 'Disable codecov')
            ->option('-s --no-scrutinizer', 'Disable scrutinizer')
            ->option('-l --no-styleci', 'Disable StyleCI')
            ->option('-L --license', 'License (m: MIT, g: GNULGPL, a: Apache2, b: BSDSimple, i: ISC, w: WTFPL)')
            ->usage($this->writer()->colorizer()->colors(''
                . '<bold>  phint init</end> <line><project></end> '
                . '<comment>--force --descr "Awesome project" --name "YourName" --email you@domain.com</end><eol/>'
                . '<bold>  phint init</end> <line><project></end> '
                . '<comment>--using laravel/lumen --namespace Project/Api --type project --license m</end><eol/>'
                . '<bold>  phint init</end> <line><project></end> '
                . '<comment>--php 7.0 --config /path/to/json --dev mockery/mockery --req adhocore/cli</end><eol/>'
            ));
    }

    /**
     * Execute the command action.
     *
     * @return void
     */
    public function execute()
    {
        $io = $this->app()->io();

        if ($using = $this->using) {
            $io->colors("Using <cyanBold>$using</end> to create project <comment>(takes some time)</end><eol/>");

            $this->_composer->createProject($this->path, $this->using);
        }

        $io->comment('Generating files ...', true);
        $this->generate($this->path, $this->values());

        $io->colors('Setting up <cyanBold>git</end><eol/>');
        $this->_git->withWorkDir($this->path)->init()->addRemote($this->username, $this->project);

        $io->colors('Setting up <cyanBold>composer</end> <comment>(takes some time)</end><eol>');
        if ($using) {
            $this->_composer->withWorkDir($this->path)->update();
        } else {
            $this->_composer->withWorkDir($this->path)->install();
        }

        $success = $this->_composer->successful();

        $success ? $io->ok('Done', true) : $io->error('Composer setup failed', true);

        $this->logging('end');
    }

    public function interact(Interactor $io)
    {
        $project = $this->project;

        if (!\preg_match('/[a-z0-9_-]/i', $project)) {
            throw new InvalidArgumentException('Project argument should only contain [a-z0-9_-]');
        }

        $io->okBold('Phint Setup', true);
        $this->logging('start');

        $this->set('path', $path = $this->prepareProjectPath());
        $this->loadConfig($this->config);

        $this->collectMissing($io);
        $this->collectPackages($io);
    }

    protected function prepareProjectPath(): string
    {
        $path = $this->project;
        $io   = $this->app()->io();

        if (!$this->_pathUtil->isAbsolute($path)) {
            $path = $this->_workDir . '/' . $path;
        }

        if (\is_dir($path)) {
            $this->projectExists($io, $path);
        } else {
            \mkdir($path, 0777, true);
        }

        return $path;
    }

    protected function projectExists(Interactor $io, string $path)
    {
        if (!$this->force && !$this->sync) {
            throw new InvalidArgumentException(
                \sprintf('Something with the name "%s" already exists!', \basename($path))
            );
        }

        if ($this->force && !$this->sync) {
            $io->error('You have set --force flag, existing files will be overwritten', true);
        }
    }

    protected function loadConfig(string $path = null)
    {
        if (empty($path)) {
            return;
        }

        if (!$this->_pathUtil->isAbsolute($path)) {
            $path = \getcwd() . '/' . $path;
        }

        if (!\is_file($path)) {
            $this->app()->io()->error('Invalid path specified for config', true);

            return;
        }

        foreach ($this->_pathUtil->readAsJson($path) as $key => $value) {
            $this->$key ?? $this->set($key, $value);
        }
    }

    protected function collectMissing(Interactor $io)
    {
        $promptConfig = [
            'type' => [
                'choices' => ['p' => 'project', 'l' => 'library', 'c' => 'composer-plugin'],
                'default' => 'l',
                'restore' => true,
            ],
            'package' => ['default' => $this->project, 'retry' => 0],
            'license' => [
                'choices' => ['m' => 'MIT', 'g' => 'GNULGPL', 'a' => 'Apache2', 'b' => 'BSDSimple', 'i' => 'ISC', 'w' => 'WTFPL'],
                'default' => 'm',
            ],
            'php' => [
                'choices' => ['5.4', '5.5', '5.6', '7.0', '7.1', '7.2', '7.3'],
                'default' => '7.0',
            ],
            'using'    => ['retry' => 0, 'extra' => ' (ENTER to skip)'],
            'keywords' => ['retry' => 0, 'extra' => ' (CSV, ENTER to skip)', 'default' => "php, {$this->project}"],
            'bin'      => ['retry' => 0, 'extra' => ' (CSV, ENTER to skip)'],

            // Donot prompt these here!
            'req'      => false,
            'dev'      => false,
            'config'   => false,
            'template' => false,
        ];

        $this->promptAll($io, $promptConfig);
    }

    protected function collectPackages(Interactor $io)
    {
        foreach (['req' => 'Required', 'dev' => 'Developer'] as $key => $label) {
            $pkgs = $this->$key ?: $this->promptPackages($label, $io);

            foreach ($pkgs as &$pkg) {
                $pkg = \strpos($pkg, ':') === false ? "{$pkg}:@stable" : $pkg;
                $pkg = \array_combine(['name', 'version'], \explode(':', $pkg, 2));
            }

            $this->set($key, $pkgs);
        }
    }

    protected function promptPackages(string $label, Interactor $io): array
    {
        $pkgs = [];

        do {
            if (!$pkg = $io->prompt($label . ' package (ENTER to skip)', null, [$this, 'validatePackage'], 0)) {
                break;
            }

            $pkgs[] = $pkg;
        } while (true);

        return $pkgs;
    }

    public function validatePackage(string $pkg): string
    {
        $pkg = \trim($pkg);

        if ($pkg && \strpos($pkg, '/') === false) {
            throw new InvalidArgumentException(
                'Package name format should be vendor/package:version (version can be omitted)'
            );
        }

        return $pkg;
    }

    protected function generate(string $projectPath, array $parameters)
    {
        $generator = new TwigGenerator($this->getTemplatePaths($parameters), $this->getCachePath());

        // Normalize license (default MIT)
        $parameters['license']   = \strtolower($parameters['license'][0] ?? 'm');
        $parameters['namespace'] = $this->makeNamespace($parameters['namespace']);
        $parameters['keywords']  = $this->makeArray($parameters['keywords'], ['php', $this->project]);
        $parameters['bin']       = $this->makeArray($parameters['bin']);

        $generator->generate($projectPath, $parameters, new CollisionHandler);
    }

    protected function makeNamespace(string $value): string
    {
        $in = new Inflector;

        $project = $this->package;
        $value   = $in->stuldyCase(\str_replace([' ', '/'], '\\', $value));
        $project = $in->stuldyCase(\str_replace([' ', '/', '\\'], '-', $project));

        if (\stripos($value, $project) === false) {
            $value .= '\\' . $project;
        }

        return $value;
    }

    protected function makeArray($value, array $default = []): array
    {
        if (empty($value)) {
            return $default;
        }

        if (\is_string($value)) {
            $value = \array_map('trim', \explode(',', $value));
        }

        return \array_values(\array_unique(\array_merge($default, $value)));
    }
}
